### PR TITLE
Modal updates: redesigned rewards modal, quantity styling/cleanup

### PR DESCRIPTION
**Pull request by anniero98**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR

- Essentially re-implement rewards modal with the updated design
- Add error message if maximum rewards applied and the '+' button clicked.
- Allow quantity modal clickable from either lineItem or displayCard
- Added `totalBalance` instead of `totalPrice` to account for negative balances (large change in logic; inline comments explain what's happening)
- Added `rewardDollarValue` and `rewardPointValue` to `constants.js`
- Updated 'totalPoints' to be `pointsEarned` and removed from state
- Styling updates for both modals (cleaner)
- Added `subtotal` and `discount` to Airtable as columns in `Transactions`; modified `ConfirmationScreen.js` to use them; and `generateConfirmationMessage` in `CheckoutScreen.js`

- Addressed belated comments from #14 (refactored some of the code in `CheckoutScreen.js`), i.e using `reduce` etc

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

### Online sources

[//]: # "Optional - copy links to any tutorial or documentation that was useful to you when working on this PR"

### Related PRs

#14 
#12 

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

## How to review

[//]: # "The order in which to review files and what to expect when testing locally"

## Next steps

[//]: # "What's NOT in this PR, doesn't work yet, and/or still needs to be done"

As for the fact that we're using `componentWillReceiveProps`: I tried to use `getDerivedStateFromProps`, but it actually is called upon local `setState` which is not what we want at all (unable to enter quantity at all...). If we keep it as is, https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html it appears that we're using Alternative 1: change _some_ of state depending on `totalBalance`. Because we have other internal state (`showError` and `modalVisible`) to be exact, I think this is fine. Also we're keeping a good amount of `draft state` around for the rewards modal. Not sure about quantity modal. I also don't know enough React to say whether we should using another pattern. Anyway, tabling this for now.

## Tests Performed, Edge Cases

[//]: # "Hopefully we will add a testing suite/CI soon, but until then note down the steps you took to test locally"

Hard to explain, but the UI interactions are complex. Removing quantity to zero should also restore any rewards that were previously applied - I chose to remove the maximum rewards possible, since otherwise customers might question why we were forcing them to "overspend" a reward. "Overspending" a reward is supported with tracking the total "balance" at all times (which could go negative). Most of the complex logic is handled in the `updateQuantityCallback`.

### Screenshots

image

modals
Included a fat gif because Github does not do videos.

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"

CC: @anniero98 @wangannie

[//]: # "This tags in both Annies as a default. Feel free to change, or add on anyone who you should be in on the conversation."

### CODE REVIEW

File: styled/modal.js [hunk 1]
```diff
@@ -1,19 +1,43 @@
...
+`;
+
+// Rewards
+export const SubheadSecondary = styled(Subhead)`
+  color: ${Colors.secondaryText};
+`;
+
+export const SubheadActive = styled(Subhead)`
+  color: ${Colors.activeText};
+`;
```
Comment (wangannie) [inline styled/modal.js]: Why not just directly use the base component?

File: components/BaseComponents.js [hunk 2]
```diff
@@ -55,6 +62,14 @@ export const ButtonLabel = styled.Text`
   color: ${props => props.color || Colors.lightest};
 `;
 
+export const BigTitleLabel = styled(BigTitle)`
+  line-height: 54px;
+  display: flex;
+  align-items: center;
```
Comment (wangannie) [inline components/BaseComponents.js]: These don't do anything since this is a child. Also, since you already have text-align, that's all you need to make it center aligned.

File: screens/ConfirmationScreen.js [hunk 3]
```diff
@@ -34,6 +35,9 @@ export default class ConfirmationScreen extends React.Component {
     if (isLoading) {
       return null;
     }
+
+    // TODO make this a value in AirTable
+    const rewardValue = 5;
     return (
       <ColumnContainer style={{ width: '100%', height: '100%', justifyContent: 'center', alignItems: 'center' }}>
         <ColumnContainer style={{ width: '33%', paddingBottom: 48 }}>
```
Comment (wangannie) [inline screens/ConfirmationScreen.js]: Sorry unrelated to this PR but can you remove Purchase Recorded (based on latest design updates)

File: components/BaseComponents.js (hunk 2 above)
Comment (wangannie) [reply components/BaseComponents.js]: Ok now I see how this was probably taken from `ButtonLabel` but that doesn't need `display: flex` or `align-items` either...we can remove both

File: screens/modals/RewardModal.js [hunk 4]
```diff
@@ -1,175 +1,198 @@
...
-    rewardStatus[i] = !rewardStatus[i];
-
-    if (apply) {
-      this.setState({
-        rewardStatus,
-        rewardsApplied: this.state.rewardsApplied + 1,
-        totalPrice: this.state.totalPrice - 5
-      });
+  updateRewardsApplied = addToApplied => {
+    if (addToApplied) {
```
Comment (wangannie) [inline screens/modals/RewardModal.js]: Very minor detail but can we make the 'Max...' notif clear immediately if the minus button is pressed? It lingers a big which could be confusing. Can we add a `this.showError(false)` or something here so the error message immediately clears if they aren't at the max?

File: styled/modal.js (hunk 1 above)
Comment (anniero98) [reply styled/modal.js]: It was really ugly in the file since it was used so often LOL

File: components/BaseComponents.js (hunk 2 above)
Comment (anniero98) [reply components/BaseComponents.js]: 👍

File: screens/ConfirmationScreen.js (hunk 3 above)
Comment (anniero98) [reply screens/ConfirmationScreen.js]: image
Designs weren't exactly up-to-date spacing-wise, so I just had it look like this for now

File: screens/CheckoutScreen.js [hunk 5]
```diff
@@ -33,98 +32,105 @@ export default class CheckoutScreen extends React.Component {
...
-      }
-    });
-    msg = msg.concat(`\nRewards Redeemed: ${this.state.rewardsApplied}\n`);
-    // Adding total price and total points earned to message. Must be called after setTotalPoints()
-    // in handleSubmit() for updated amount.
-    msg = msg.concat(`\nTotal Price: $${this.state.totalPrice.toFixed(2)}\n`);
-    msg = msg.concat(`Total Points Earned: ${totalPoints}`);
-    return msg;
-  };
+    // Calculate actual discount and sale; handle negative balances
```
Comment (anniero98) [inline screens/CheckoutScreen.js]: @thumn This is where subtotal gets calculated!

Comment (thumn) [reply screens/CheckoutScreen.js]: ty!

File: screens/CheckoutScreen.js [hunk 6]
```diff
@@ -33,98 +32,105 @@ export default class CheckoutScreen extends React.Component {
...
+      (points, lineItem) => points + lineItem.points * lineItem.quantity,
+      0
+    );
+    // Customer cannot earn points with rewards dollars; assumes a reward's point multiplier per dollar is 100 pts
+    pointsEarned -= this.state.rewardsApplied * rewardDollarValue * 100;
+
+    // TODO this might be a design edge case now that rewards applied can bring the technical balance to a negative
+    if (pointsEarned < 0) {
+      if (this.state.totalBalance > 0) {
+        console.log(
```
Comment (thumn) [inline screens/CheckoutScreen.js]: Would this be more helpful as an error message for us to debug :0 we would remove this in the MVP right

Comment (anniero98) [reply screens/CheckoutScreen.js]: Good pt!! I left it as a log because of the low likelihood but possible legal scenario in which items' multipliers are low ;o 

But ya I think we should figure out a totally diff logging system for prod anyway! (I don't know where console.logs go LOL?) will leave it for now 😅

File: screens/ConfirmationScreen.js [hunk 7]
```diff
@@ -34,32 +35,35 @@ export default class ConfirmationScreen extends React.Component {
...
+              <Body>Points Earned</Body>
+              <Body>{transaction.pointsEarned} pts</Body>
+            </SpaceBetweenRowContainer>
+            <SpaceBetweenRowContainer style={{ paddingBottom: 12 }}>
+              <Body>Rewards Redeemed</Body>
+              <Body>{transaction.rewardsApplied}</Body>
+            </SpaceBetweenRowContainer>
+            <SpaceBetweenRowContainer>
+              <Body>Subtotal</Body>
+              <Body>{displayDollarValue(transaction.discount)}</Body>
```
Comment (thumn) [inline screens/ConfirmationScreen.js]: here!
